File: src/vehicle_drivers/vision_node/src/yolo_video_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"
#include "opencv2/opencv.hpp"
#include "opencv2/dnn.hpp"
#include "image_transport/image_transport.hpp"

class YoloVideoSubscriber : public rclcpp::Node
{
public:
    YoloVideoSubscriber()
        : Node("yolo_video_subscriber")
    {
        // 使用 image_transport 订阅，自动解压
        image_transport_ = std::make_shared<image_transport::ImageTransport>(shared_from_this());
        sub_ = image_transport_->subscribe(
            "/camera/image_raw", 10,
            std::bind(&YoloVideoSubscriber::imageCallback, this, std::placeholders::_1));

        // 加载 YOLO 模型（请确保模型文件放在工作目录或指定绝对路径）
        std::string model_path = "yolo11n.pt";
        net_ = cv::dnn::readNet(model_path);
        if (net_.empty()) {
            RCLCPP_ERROR(this->get_logger(), "Failed to load YOLO model from %s", model_path.c_str());
        }
    }

private:
    void imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr & msg)
    {
        try {
            // 解码后的图像已经是 bgr8
            cv::Mat frame = cv_bridge::toCvShare(msg, "bgr8")->image;

            // YOLO 推理
            cv::Mat blob = cv::dnn::blobFromImage(frame, 1.0/255.0, cv::Size(640,640), cv::Scalar(), true, false);
            net_.setInput(blob);

            std::vector<cv::Mat> outputs;
            net_.forward(outputs, net_.getUnconnectedOutLayersNames());

            for (const auto & output : outputs) {
                auto *data = (float *)output.data;
                for (int i = 0; i < output.rows; ++i, data += output.cols) {
                    float confidence = data[4];
                    if (confidence > 0.5) {
                        int x = static_cast<int>(data[0] * frame.cols);
                        int y = static_cast<int>(data[1] * frame.rows);
                        int w = static_cast<int>(data[2] * frame.cols);
                        int h = static_cast<int>(data[3] * frame.rows);
                        cv::rectangle(frame, cv::Rect(x, y, w, h), cv::Scalar(0,255,0), 2);
                    }
                }
            }

            cv::imshow("YOLO Detection", frame);
            cv::waitKey(1);
        } catch (const std::exception & e) {
            RCLCPP_ERROR(this->get_logger(), "Error processing image: %s", e.what());
        }
    }

    std::shared_ptr<image_transport::ImageTransport> image_transport_;
    image_transport::Subscriber sub_;
    cv::dnn::Net net_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<YoloVideoSubscriber>());
    rclcpp::shutdown();
    return 0;
}

File: src/vehicle_drivers/vision_node/vision_package/yolo_video_subscriber.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2


class YoloVideoSubscriber(Node):
    def __init__(self):
        super().__init__('yolo_video_subscriber')
        self.bridge = CvBridge()

        self.subscription = self.create_subscription(
            Image, '/camera/image_raw', self.image_callback, 10)

        # 需要 ONNX 格式模型，例如通过 `yolo export model=yolo11n.pt format=onnx` 导出
        model_path = 'yolo11n.onnx'
        self.net = cv2.dnn.readNet(model_path)
        if self.net.empty():
            self.get_logger().error(f'Failed to load YOLO model from {model_path}')

    def image_callback(self, msg):
        try:
            frame = self.bridge.imgmsg_to_cv2(msg, 'bgr8')

            blob = cv2.dnn.blobFromImage(
                frame, 1.0 / 255.0, (640, 640), swapRB=True, crop=False)
            self.net.setInput(blob)
            outputs = self.net.forward(self.net.getUnconnectedOutLayersNames())

            for output in outputs:
                for detection in output:
                    confidence = detection[4]
                    if confidence > 0.5:
                        x = int(detection[0] * frame.shape[1])
                        y = int(detection[1] * frame.shape[0])
                        w = int(detection[2] * frame.shape[1])
                        h = int(detection[3] * frame.shape[0])
                        cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)

            cv2.imshow('YOLO Detection', frame)
            cv2.waitKey(1)
        except Exception as e:
            self.get_logger().error(f'Error processing image: {e}')


def main(args=None):
    rclpy.init(args=args)
    node = YoloVideoSubscriber()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
